dex/price-discovery/src/user_actions/redeem.rs

```diff
@@ -0,0 +1,104 @@
+use crate::user_actions::user_deposit_withdraw::INVALID_PAYMENT_ERR_MSG;
+
+multiversx_sc::imports!();
+
+#[multiversx_sc::module]
+pub trait RedeemModule:
+    crate::common_storage::CommonStorageModule
+    + crate::events::EventsModule
+    + crate::phase::PhaseModule
+    + crate::redeem_token::RedeemTokenModule
+    + multiversx_sc_modules::default_issue_callbacks::DefaultIssueCallbacksModule
+{
+    /// After all phases have ended,
+    /// users can withdraw their fair share of launched tokens,
+    /// while the owner can withdraw the accepted tokens.
+    #[payable("*")]
+    #[endpoint]
+    fn redeem(&self) -> EgldOrEsdtTokenPayment {
+        let phase = self.get_current_phase();
+        self.require_redeem_allowed(&phase);
+
+        let caller = self.blockchain().get_caller();
+        let owner = self.owner_address().get();
+        if caller == owner {
+            let redeemed_tokens = self.owner_redeem(&caller);
+
+            self.emit_redeem_event(
+                None,
+                &BigUint::zero(),
+                &redeemed_tokens.token_identifier,
+                &redeemed_tokens.amount,
+            );
+
+            return redeemed_tokens;
+        }
+
+        let (payment_token, payment_amount) = self.call_value().single_fungible_esdt();
+        let bought_tokens = self.user_redeem(&caller, &payment_token, &payment_amount);
+
+        self.emit_redeem_event(
+            Some(&payment_token),
+            &payment_amount,
+            &bought_tokens.token_identifier,
+            &bought_tokens.amount,
+        );
+
+        bought_tokens
+    }
+
+    fn owner_redeem(&self, owner: &ManagedAddress) -> EgldOrEsdtTokenPayment {
+        let launched_token_supply = self.launched_token_balance().get();
+        require!(
+            launched_token_supply > 0,
+            "May not withdraw tokens as launched tokens were not deposited"
+        );
+
+        let accepted_token_id = self.accepted_token_id().get();
+        let accepted_token_balance = self.accepted_token_balance().get();
+        self.send()
+            .direct(owner, &accepted_token_id, 0, &accepted_token_balance);
+
+        EgldOrEsdtTokenPayment::new(accepted_token_id, 0, accepted_token_balance)
+    }
+
+    fn user_redeem(
+        &self,
+        user: &ManagedAddress,
+        payment_token: &TokenIdentifier,
+        payment_amount: &BigUint,
+    ) -> EgldOrEsdtTokenPayment {
+        let redeem_token_id = self.redeem_token().get_token_id();
+        require!(payment_token == &redeem_token_id, INVALID_PAYMENT_ERR_MSG);
+
+        let launched_token_supply = self.launched_token_balance().get();
+        if launched_token_supply == 0 {
```

Comment:
How could you get to this situation? What if launched_token_supply is > 0 but < than the SC balance?

Reply:
If the owner doesn't deposit any tokens. I'm not sure I get the second question? How would you ever reach such a situation?